### ISSUE TITLE
workgroup_1_2024_I

### ISSUE THREAD

**Issue by alexanderquispe** [d2cml-ai/CausalAI-Course#828]
# Workgroup 1 

## Math (5 points)
1. Make the proof of the Frisch–Waugh–Lovell theorem using Matrix Algebra. Let the following equation

$$\overset{nx1}y=\overset{nxk_1}D \overset{k_1x1}\beta_1 + \overset{nxk_2}W \overset{k_2x1}\beta_2+\overset{nx1}\mu$$ 

The Frisch-Waugh-Lovell Theorem states that we can obtain $\beta_1$ through the alternative regression:

$$\epsilon^y = \epsilon^D \psi+\xi $$

show that $$\psi = \beta_1$$

2. Show that the Conditional Expectation Function minimizes expected squared error. You can start with the following equations. 
$$Y = m(X) + \epsilon  $$
$$m(X) = \underset{g(X)}{\text{arg min}} \, E[(Y - g(X))^2] $$
$$E[(Y - g(X))^2] = E[(Y - m(X) + m(X) - g(X))^2] $$

you have to show that 
$$E[(Y - g(X))^2] = E[(\epsilon)^2] $$

You will have to use law of iterated/total expectations (LIE)

## Replication 1  - Code (8 Points)
You will have to replicate the second lab. 

* PM1_Notebook_Inference.ipynb (Python)
* pm1-notebook-inference.ipynb (R)
* pm1-notebook-inference.ipynb (Julia)

But you have to follow the next instructions:
1. Focus on the subset of college-advanced-educated workers (**scl, clg, ad** variables). The analysis should be analogous to what we’ve presented – explaining basic, control and partialling out model, generating point estimates and standard errors. hint keep if scl == 1 | clg == 1 | ad == 1 

2. Use appropiate plots (i.e hist, barplots, scatter plots , pie plots, etc) to describe main varaibles (wage, log-wage, sex, some college, college graduate, avdanced degree, Experience)

* Turorial plot for Python 
Use this material (https://github.com/alexanderquispe/Diplomado_PUCP/tree/main/Lecture_9)

* Tutorial plots for R
https://www.kaggle.com/code/benroshan/part-4-data-visualization-with-r/notebook

* Tutorial plots for Julia
https://www.analyticsvidhya.com/blog/2021/05/data-visualizations-in-julia-using-plots-jl-with-practical-implementation/

4. Plot the confidence Interval of sex's coefficient for a different models (basic, control, and partially out). All three coefficients must be in one figure. Explain what you find.  
Use these references:
Coefplot in R:
https://www.r-statistics.com/2010/07/visualization-of-regression-coefficients-in-r/
Coefplot in Julia:
https://discourse.julialang.org/t/plotting-coefplot/45779/4
Coefplot in Python:
https://zhiyzuo.github.io/Python-Plot-Regression-Coefficient/

5. You will also include a replication of the next figure for both groups, female and male :
Check this [paper](https://eml.berkeley.edu/~cle/wp/wp62.pdf) for help. You may follow the instructions about  the regressions to create the figures.  
![image](https://user-images.githubusercontent.com/27368860/132098220-e46c9301-77f8-4845-aee2-fdf7e9132469.png)
Help: Use the Log(wage) variable and the experience variables. 

You will have only two plotted lines (Actual/Predicted(fitted)) for these College-educated workers. You have to create two separate figures, one for female and one for male. Could you explain the different patterns that you find? 

### Objective
To understand and manually implement the process of k-fold cross-validation for selecting the optimal regularization parameter (alpha) in Lasso Regression.

#### 1. Data Preparation
Load the March Supplement of the U.S. Current Population Survey, year 2015. (*wage2015_subsample_inference.Rdata*)

#### 2. Define a Range of Alpha (Lambda in our equation) Values
Create a list or array of alpha values to iterate over. These will be the different regularization strengths you will test. You should use at least 5 different values of alpha. Hint: You can use small numbers from 0.1 to 0.5

#### 3. Partition the Dataset for k-Fold Cross-Validation
Divide the dataset into `k` subsets (or folds). If you are working with a regression task, ensure the data is shuffled. For classification tasks, the folds should be stratified.

#### 4. Lasso Regression Implementation
Implement a function to fit a Lasso Regression model given a training dataset and an alpha value. The function should return the model's coefficients and intercept.

#### 5. Cross-Validation Loop
For each alpha value in your predefined list:
   - Initialize an empty list to keep track of the Mean Squared Errors (MSE) for each fold.
   - Perform k-fold cross-validation:
     - In each iteration, designate one fold as the validation set, and the combination of the remaining folds as the training set.
     - Fit the Lasso model on the training set.
     - Predict and evaluate against the validation set.
     - Append the MSE for this fold to the list of MSEs.
   - Calculate and store the average MSE over all k folds for this alpha.

#### 6. Selection of Optimal Alpha
After testing all alphas, select the alpha that yielded the lowest average MSE in cross-validation.

#### 7. Model Training and Evaluation
Using the optimal alpha value, train a final Lasso model on the entire training set. Evaluate its performance on a separate test set to assess its generalizability.

#### 8. Report Results
Report the selected alpha value and the performance of the final model on the test set. Discuss the implications of your findings.

### Expected Output
- A plot of the cross-validated MSE for each alpha value.
- The value of the optimal alpha.
- The test set MSE of the final Lasso model trained with the optimal alpha.

# Settings 

- Create only one branch for your group: **_group1_lab1_** or  **_group2_lab1_**
- Name the scripts as  **"group1_lab1_python"**, **"group1_lab1_julia"** & **"group1_lab1_r"**
- Do by yourself the **merge** and **Delete your branch** once you are done. 
- Save both files in [CausalAI-Course/Labs/replication_1/](https://github.com/alexanderquispe/CausalAI-Course/tree/main/labs/replication_1)
- Deadline: Deadline: 13 April (23:59)**
